### PR TITLE
[Rust] add substrait for flink and be compatible for other engines

### PR DESCRIPTION
**Pull request by mag1c1an1**
add flink expression to substrait

add more functions

add more tests

add base schema for namedscan, substriat type to arrow type

compatibility

switch to java8

### CODE REVIEW

File: lakesoul-flink/src/main/java/org/apache/flink/lakesoul/source/LakeSoulLookupTableSource.java [hunk 1]
```diff
@@ -235,6 +235,7 @@ public DynamicTableSource copy() {
         lsts.projectedFields = this.projectedFields;
         lsts.remainingPartitions = this.remainingPartitions;
         lsts.filter = this.filter;
+        lsts.filter = this.filter;
```
Comment (Ceng23333) [inline lakesoul-flink/src/main/java/org/apache/flink/lakesoul/source/LakeSoulLookupTableSource.java:238]: duplicate code

File: lakesoul-flink/src/main/java/org/apache/flink/lakesoul/source/LakeSoulSource.java [hunk 2]
```diff
@@ -52,15 +55,18 @@ public LakeSoulSource(TableId tableId,
                           List<String> pkColumns,
                           Map<String, String> optionParams,
                           @Nullable List<Map<String, String>> remainingPartitions,
-                          @Nullable FilterPredicate filter) {
+                          @Nullable FilterPredicate filterStr,
```
Comment (Ceng23333) [inline lakesoul-flink/src/main/java/org/apache/flink/lakesoul/source/LakeSoulSource.java:58]: improper name 'filterStr'

File: lakesoul-flink/src/main/java/org/apache/flink/lakesoul/source/LakeSoulOneSplitRecordsReader.java [hunk 3]
```diff
@@ -129,7 +133,11 @@ private void initializeReader() throws IOException {
         }
 
         if (filter != null) {
```
Comment (Ceng23333) [inline lakesoul-flink/src/main/java/org/apache/flink/lakesoul/source/LakeSoulOneSplitRecordsReader.java:135]: Will two kinds of filter cause conflict?

File: lakesoul-flink/src/main/java/org/apache/flink/lakesoul/substrait/SubstraitUtil.java [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+                } else {
+                    last = e;
+                }
+            }
+        }
+        Plan filter = toFilter(last, tableName, arrowSchema);
+        return Tuple2.of(SupportsFilterPushDown.Result.of(accepted, remaining), planToProto(filter));
+    }
+
+    static Schema toArrowSchema(String tableSchema) {
```
Comment (Ceng23333) [inline lakesoul-flink/src/main/java/org/apache/flink/lakesoul/substrait/SubstraitUtil.java:78]: No diff from Schema.fromJSON

File: lakesoul-flink/src/main/java/org/apache/flink/lakesoul/substrait/SubstraitUtil.java [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+import org.apache.flink.table.expressions.ResolvedExpression;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class SubstraitUtil {
```
Comment (Ceng23333) [inline lakesoul-flink/src/main/java/org/apache/flink/lakesoul/substrait/SubstraitUtil.java:34]: Move to lakesoul-io-java maybe better.

File: native-io/lakesoul-io-java/src/main/java/com/dmetasoul/lakesoul/lakesoul/io/NativeIOReader.java [hunk 6]
```diff
@@ -34,6 +36,23 @@ public void addFilter(String filter) {
         ioConfigBuilder = libLakeSoulIO.lakesoul_config_builder_add_filter(ioConfigBuilder, filter);
     }
 
+    /**
+     * usually use only once
+     *
+     * @param plan Filter{}
+     */
+    public void addFilterProto(Plan plan) {
```
Comment (Ceng23333) [inline native-io/lakesoul-io-java/src/main/java/com/dmetasoul/lakesoul/lakesoul/io/NativeIOReader.java:44]: refer com.dmetasoul.lakesoul.meta.jnr.NativeMetadataJavaClient#executeInsert

File: rust/lakesoul-io/src/datasource/physical_plan/merge.rs [hunk 7]
```diff
@@ -255,20 +255,36 @@ pub async fn prune_filter_and_execute(
     df: DataFrame,
     request_schema: SchemaRef,
     filter_str: Vec<String>,
```
Comment (Ceng23333) [inline rust/lakesoul-io/src/datasource/physical_plan/merge.rs:257]: use Vec<Expr> as input directly

File: lakesoul-flink/src/main/java/org/apache/flink/lakesoul/source/LakeSoulOneSplitRecordsReader.java (hunk 3 above)
Comment (mag1c1an1) [reply lakesoul-flink/src/main/java/org/apache/flink/lakesoul/source/LakeSoulOneSplitRecordsReader.java:135]: No, `filterSter` is null forever in above code. It is used to debug the difference of result datafusion::expr of two kinds filters.

**Comment by mag1c1an1** [comment]
rebase from main, please review the dependency

**Review by Ceng23333 — COMMENTED** [review]
Please provide some flink test cases.

File: lakesoul-flink/src/test/java/org/apache/flink/lakesoul/test/SubstraitTest.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+import java.util.Comparator;
+import java.util.List;
+import java.util.TimeZone;
+import java.util.concurrent.ExecutionException;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class SubstraitTest extends AbstractTestBase {
+
+    private final String BATCH_TYPE = "batch";
```
Comment (Ceng23333) [inline lakesoul-flink/src/test/java/org/apache/flink/lakesoul/test/SubstraitTest.java:19]: Add tests which filter on hash column/range column

**Review by Ceng23333 — APPROVED** [review]

**Pull request merged**